### PR TITLE
FIX: Replace very inefficient discrete _get_trial

### PR DESCRIPTION
**Pull request by KatharineShapcott**
On my data x20 speedup

Author Guidelines
-----------------
- [x] Is the change set **< 600 lines**?
- [x] Was the code checked for memory leaks/performance bottlenecks?
- [ ] Is the code running locally **and** on the ESI cluster?
- [ ] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [ ] Do objects in the global package namespace perform proper parsing of their input? 
- [ ] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [ ] Are all docstrings complete and accurate?
- [ ] Is the CHANGELOG.md up to date?

### CODE REVIEW

**Comment by KatharineShapcott** [comment]
My assumption is that most spike/event data will be sorted like ours is and therefore only part of the file needs to be loaded from disk, hence the dramatic speedup. But it doesn't require sorted data to work.

File: syncopy/datatype/discrete_data.py [hunk 1]
```diff
@@ -184,7 +185,12 @@ def trialtime(self):
 
     # Helper function that grabs a single trial
     def _get_trial(self, trialno):
-        return self._data[self.trialid == trialno, :]
+        this_trl = self.trialid == trialno
+        if not np.any(this_trl):
+            return self._data[None, :]
```
Comment (tensionhead) [inline syncopy/datatype/discrete_data.py:190]: Isn't this the same as 
```python
return self._data[()]
```
and why would we want to return the whole dataset in case the `trialno` does no exist?

Comment (KatharineShapcott) [reply syncopy/datatype/discrete_data.py:190]: Ah sorry I want to return an empty slice e.g. self._data[0:0, :]

**Comment by KatharineShapcott** [comment]
While we're at it, I noticed that printing is very slow (although a lot faster after this speedup) since many different attributes are actually calling `_get_trial` under the hood. 

```
sample
817 ms ± 6.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
time
401 ms ± 5.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
trials
384 ms ± 5.05 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
trialtime
496 ms ± 4.76 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
trialinfo
392 ms ± 14.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

What if I make a `_get_trial` specifically for the SpikeData class that I can assume is sorted. Then storing a searchsorted call in a private variable `_trialslice` would be very fast e.g.

```
def _get_trial(self, trialno):
        if not hasattr(self, '_trialslice'):
            self._trialslice = np.searchsorted(self._data[:,self.dimord.index("sample")], self.sampleinfo.ravel())
            self._trialslice = self._trialslice.reshape(self.sampleinfo.shape)
        idx = slice(self._trialslice[trialno,0], self._trialslice[trialno,1])
        return self._data[idx,:]
```

**Comment by tensionhead** [comment]
Puh, I am not sure I can follow you here.. so when I do this for some synthetic data `spd` (`syncopy.tests.synth_data.poisson_noise`) I have:
```python
In [19]: spd.sampleinfo
Out[19]: 
array([[      0.,  180609.],
       [ 200000.,  396256.],
       [ 400000.,  582885.],
       [ 600000.,  783863.],
       [ 800000.,  996461.],
       [1000000., 1184701.],
       [1200000., 1392257.],
       [1400000., 1592449.],
       [1600000., 1783830.],
       [1800000., 1999999.]])
```
and after your proposed operation:
```python
In [17]: trialslice.reshape(spd.sampleinfo.shape)
Out[17]: 
array([[     0,  17986],
       [ 19928,  39816],
       [ 40191,  58440],
       [ 60106,  78473],
       [ 80111,  99810],
       [100169, 118774],
       [120309, 139543],
       [140340, 159565],
       [160378, 178619],
       [180300, 200000]])
```
which I am not sure what it is exactly, looks like sub-intervals of the 1st trial?!

**Comment by KatharineShapcott** [comment]
Yes definitely confusing compared to continuous data... trialslice is not samples anymore but is in units of number of spikes. So what this tells me is that your first trial contains 17,986 spikes, while your second trial contains 19,888 spikes and the first spike that falls within that trial is at index 19,928 etc. 

The problem with this solution is that if you change the trialdefinition or use selectdata you need to update trialslice at the same time. Are you around for a meeting about this at some point?

**Comment by KatharineShapcott** [comment]
Okay I think I get why this crashed now. With the previous version of .trials it was impossible to return an empty array, it would just be skipped over. So when you tried to select trials[0] here you would really get trials[1]. That's fine but the problem seems to be that if you try to use selectdata to select that empty trial and there is no data at all then everything gets set to None including the trialdefinition. 

```
spikes.trials[0]
Out[36]: array([], shape=(0, 3), dtype=int64)

In [37]: spikes.trialdefinition
Out[37]: 
array([[   0.,   10.,    0.],
       [  10.,  100.,    0.],
       [ 200.,  500.,    0.],
       [ 600.,  700.,    0.],
       [ 800., 1000.,    0.]])

In [38]: selected.trialdefinition
Out[38]: array(None, dtype=object)
```

Which causes the following error

```
selected.trials[0]

SyNCoPy encountered an error in 

/gs/home/shapcottk/.conda/envs/syncopyoe/lib/python3.8/site-packages/IPython/core/interactiveshell.py, line 3361 in run_code
        exec(code_obj, self.user_global_ns, self.user_ns)

--------------------------------------------------------------------------------
Abbreviated traceback:

<ipython-input-40-391b5f5f858d>, line 1 in <cell line: 1>
        selected.trials[0]

Use `import traceback; import sys; traceback.print_tb(sys.last_traceback)` for full error traceback.

TypeError: 'NoneType' object is not subscriptable
```

**Comment by tensionhead** [comment]
Mmh, so in the tests we select empty trials, that's somewhat surprising! I would consider this an edge case.. There's loads of CI output so I can't really see what was happening though. Shall I look into this or you think you can get it to work?

**Comment by KatharineShapcott** [comment]
Yeah it's kind of an edge case but since the test data contains very few spikes it occurs quite often in the tests I think. 

I tried to avoid it by returning the empty arrays instead of None, let's see what happens now.

**Comment by tensionhead** [comment]
Mmh, we just deprecated `slice` as valid selection parameter with #407 , as they are not serializable and hence the `cfg` can't be saved to json. But you branched out befor that got merged today no?!

**Comment by tensionhead** [comment]
ok, I will try to merge dev into your branch :crossed_fingers:

**Comment by tensionhead** [comment]
So `dev` itself got repaired, can't really tell atm what is happening here tbh, maybe @dfsp-spirit got an idea?!

**Comment by KatharineShapcott** [comment]
I don't think we need to revert removing the unique from sample, the tests passed from that change before:

image

**Comment by KatharineShapcott** [comment]
@tensionhead The one that's failing is a test_ed_dataselection within the test_discretedata. Can we simply remove that test? 

https://github.com/esi-neuroscience/syncopy/blob/ba57fd5248e022bd1fea89c5a7695a685b235ea0/syncopy/tests/test_discretedata.py#L564

**Comment by KatharineShapcott** [comment]
Maybe it's failing because we switched away from this relative indexing?

**Comment by tensionhead** [comment]
Phew.. right we also have selection tests witin the individual data class tests :sweat_smile: Yup, I think it is safe to remove, this whole `EventData` probably needs an overhaul anyways.. then event id indexing is so weird.

File: syncopy/datatype/methods/definetrial.py [hunk 2]
```diff
@@ -335,20 +335,13 @@ def definetrial(obj, trialdefinition=None, pre=None, post=None, start=None,
...
-        # no data - empty object, can happen due to a selection
-        else:
-            tgt.trialid = None
-            tgt._trialdefinition = None
+        idx = np.searchsorted(samples, tgt.sampleinfo.ravel())
+        idx = idx.reshape(tgt.sampleinfo.shape)
+
+        tgt._trialslice = [slice(st,end) for st,end in idx]
+        tgt.trialid = np.full((samples.shape), -1, dtype=int)
+        for itrl, itrl_slice in enumerate(tgt._trialslice):
```
Comment (tensionhead) [inline syncopy/datatype/methods/definetrial.py:343]: Nice :+1: :slightly_smiling_face:

**Review by tensionhead — COMMENTED** [review]
Looks good, nice small surgical changes! I just have 1 minor question, see below.

EDIT: Looks like the original failing test just evaporated with the re-write of the `test_selectdata.py` :slightly_smiling_face:

File: syncopy/datatype/discrete_data.py [hunk 3]
```diff
@@ -156,6 +158,9 @@ def trialid(self, trlid):
             print("SyNCoPy core - trialid: Cannot assign `trialid` without data. " +
                   "Please assing data first")
             return
+        if (self.data.shape[0] == 0) and (trlid.shape[0] == 0):
+            self._trialid = np.array(trlid, dtype=int)
```
[the remark this answers is not in the document]
Comment (KatharineShapcott) [reply syncopy/datatype/discrete_data.py:162]: Yes exactly, but it used to happen in the tests.

File: syncopy/datatype/discrete_data.py [hunk 4]
```diff
@@ -115,6 +115,8 @@ def sample(self):
         """Indices of all recorded samples"""
         if self.data is None:
             return None
+        # return self.data[:, self.dimord.index("sample")]
+        # there should be only one event per sample number?!
         return np.unique(self.data[:, self.dimord.index("sample")])
```
[the remark this answers is not in the document]
Comment (KatharineShapcott) [reply syncopy/datatype/discrete_data.py:120]: I think we can remove the unique. I don't understand why it was ever there.

File: syncopy/datatype/discrete_data.py [hunk 5]
```diff
@@ -115,6 +115,8 @@ def sample(self):
         """Indices of all recorded samples"""
         if self.data is None:
             return None
+        # return self.data[:, self.dimord.index("sample")]
+        # there should be only one event per sample number?!
```
[the remark this answers is not in the document]
Comment (KatharineShapcott) [reply syncopy/datatype/discrete_data.py:119]: Not true for spike data, you quite often get multiple spikes per sample.

**Comment by KatharineShapcott** [comment]
@tensionhead Should I merge in the other speedups?

File: syncopy/datatype/discrete_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/datatype/discrete_data.py:160]: of course.. :sweat_smile:

**Comment by tensionhead** [comment]
Yeah was wondering myself, but if the next CI run passes I think we are good

**Review by tensionhead — APPROVED** [review]
Great, together with #415 we hopefully made significant perfomance gains for `DiscreteData`
